data/transactions/logic/resources.go

```diff
@@ -0,0 +1,328 @@
+	r.shareAccountAndHolding(hdr.Sender, id)
+	r.shareAccountAndHolding(tx.AssetReceiver, id)
+
+	if !tx.AssetSender.IsZero() {
+		r.shareAccountAndHolding(tx.AssetSender, id)
+	}
+
+	if !tx.AssetCloseTo.IsZero() {
+		r.shareAccountAndHolding(tx.AssetCloseTo, id)
+	}
+}
+
+// allowsHolding checks if a holding is available under the txgroup sharing rules
+func (r *resources) allowsHolding(addr basics.Address, ai basics.AssetIndex) bool {
+	if _, ok := r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: ai}]; ok {
+		return true
+	}
+	// All holdings of created assets are available
+	for _, created := range r.createdAsas {
+		if created == ai {
+			return true
+		}
+	}
+	return false
+}
+
+// allowsLocals checks if a local state is available under the txgroup sharing rules
+func (r *resources) allowsLocals(addr basics.Address, ai basics.AppIndex) bool {
+	if _, ok := r.sharedLocals[ledgercore.AccountApp{Address: addr, App: ai}]; ok {
+		return true
+	}
+	// All locals of created apps are available
+	for _, created := range r.createdApps {
+		if created == ai {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *resources) requireHolding(acct basics.Address, id basics.AssetIndex) error {
+	/* Previous versions allowed inner appls with zeros in "required" places,
+	   even if that 0 resource should have be inaccessible, because the check
+	   was done at itxn_field time, and maybe the app siimply didn't set the
+	   field. */
+	if id == 0 || acct.IsZero() {
+		return nil
+	}
+	if !r.allowsHolding(acct, id) {
+		return fmt.Errorf("invalid Holding access %s x %d", acct, id)
+	}
+	return nil
+}
+
+func (r *resources) requireLocals(acct basics.Address, id basics.AppIndex) error {
+	if !r.allowsLocals(acct, id) {
+		return fmt.Errorf("invalid Local State access %s x %d", acct, id)
+	}
+	return nil
+}
+
+func (r *resources) allowsAssetTransfer(hdr *transactions.Header, tx *transactions.AssetTransferTxnFields) error {
+	err := r.requireHolding(hdr.Sender, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer Sender: %v", err)
+	}
+	err = r.requireHolding(tx.AssetReceiver, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer AssetReceiver: %v", err)
+	}
+	err = r.requireHolding(tx.AssetSender, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer AssetSender: %v", err)
+	}
+	err = r.requireHolding(tx.AssetCloseTo, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer AssetCloseTo: %v", err)
+	}
+	return nil
+}
+
+func (r *resources) fillAssetFreeze(hdr *transactions.Header, tx *transactions.AssetFreezeTxnFields) {
+	r.sharedAccounts[hdr.Sender] = struct{}{}
+	id := tx.FreezeAsset
+	r.sharedAsas[id] = struct{}{}
+	r.shareAccountAndHolding(tx.FreezeAccount, id)
+}
+
+func (r *resources) allowsAssetFreeze(hdr *transactions.Header, tx *transactions.AssetFreezeTxnFields) error {
+	err := r.requireHolding(tx.FreezeAccount, tx.FreezeAsset)
+	if err != nil {
+		return fmt.Errorf("afrz FreezeAccount: %v", err)
```

Comment:
```suggestion
		return fmt.Errorf("afrz FreezeAccount: %w", err)
```
This file uses `%v` in many places to insert an error, but `%w` would properly wrap it

Reply:
Thanks, you're correct, of course.